src/coreclr/jit/emit.cpp

```diff
@@ -8208,76 +8208,92 @@ CORINFO_FIELD_HANDLE emitter::emitSimdConst(simd_t* constValue, emitAttr attr)
 //
 void emitter::emitSimdConstCompressedLoad(simd_t* constValue, emitAttr attr, regNumber targetReg)
 {
-    assert(EA_SIZE(attr) >= 8);
-    unsigned    cnsSize = EA_SIZE(attr);
-    instruction ins     = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
+    assert(EA_SIZE(attr) >= 8 && EA_SIZE(attr) <= 64);
 
-    for (unsigned size = 4; size < cnsSize; size *= 2)
+    unsigned    cnsSize  = EA_SIZE(attr);
+    unsigned    dataSize = cnsSize;
+    instruction ins      = (cnsSize == 8) ? INS_movsd_simd : INS_movups;
+
+    // Most constant vectors tend to have repeated values, so we will first check to see if
+    // we can replace a full vector load with a smaller broadcast.
+
+    if ((dataSize == 64) && (constValue->v256[1] == constValue->v256[0]))
     {
-        // If the constant has zero upper elements/lanes, we can use a smaller load,
-        // because all scalar and vector loads from memory zero the upper.
+        assert(emitComp->IsBaselineVector512IsaSupportedDebugOnly());
+        dataSize = 32;
+        ins      = INS_vbroadcastf32x8;
```

Comment:
nit: It's not as important nowadays and I wouldn't block the PR on it. But it might be beneficial to take a `simdBaseType` and use it to pick a more "optimal" instruction where possible.

This "can" impact alignment, atomicity, general readability of the disasm output, potentially even performance, etc. So trying to match the broadcast used to the user type is beneficial where possible. Of course we may not always have a type and `float` is the common fallback, but doing a best case of picking `f32x8` vs `f64x4` vs `i32x8` vs `i64x4` and so on is beneficial.

Reply:
This is what we chatted about yesterday.  This method is only called from `genSetRegToConst`, which generates code for uncontained `CNS_VEC`, because its consumer needs it in a register.  `CNS_VEC` doesn't have a `simdBaseType` and apparently doesn't have room to add one.

All we know at this callsite is what the constant is and what register to put it in. The best we can do is produce a register with the correct content, hopefully in the most efficient way possible.

Reply:
You should still be able to get the user of the node to see what the likely type is and default to float if a user doesn’t exist or if it isn’t a hwintrinsic node.

Reply:
True, I could make a semi-informed guess, but in this particular case, it's a purely aesthetic concern. Even going back to old CPU architectures that had a bypass delay when switching between integer and float domains, load and store instructions have always been type agnostic in practice. And element size doesn't matter here because we're not using embedded masking --otherwise the constant would be contained in another instruction and not getting its own codegen.

So I think in this case, guessing the type and changing the instruction mnemonic is unnecessary complexity.